python-frontend/src/main/java/org/sonar/python/semantic/v2/types/Propagation.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.semantic.v2.types;
+
+import java.util.HashSet;
+import java.util.Set;
+import org.sonar.plugins.python.api.tree.Expression;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.python.semantic.v2.SymbolV2;
+import org.sonar.python.semantic.v2.SymbolV2Utils;
+import org.sonar.python.semantic.v2.UsageV2;
+import org.sonar.python.tree.NameImpl;
+import org.sonar.python.types.v2.PythonType;
+import org.sonar.python.types.v2.UnionType;
+
+public abstract class Propagation {
+
+  final Set<SymbolV2> variableDependencies = new HashSet<>();
+  final Set<Propagation> dependents = new HashSet<>();
+
+  final SymbolV2 lhsSymbol;
+  final Name lhsName;
+
+  protected Propagation(SymbolV2 lhsSymbol, Name lhsName) {
+    this.lhsSymbol = lhsSymbol;
+    this.lhsName = lhsName;
+  }
+
+  /**
+   * This is used for AST-based type inference in try/catch statements
+   * @return true if the propagation effectively changed the inferred type of assignment LHS
+   */
+  boolean propagate(Set<SymbolV2> initializedVars) {
+    PythonType rhsType = rhsType();
+    if (initializedVars.add(lhsSymbol)) {
+      lhsSymbol.usages().stream()
+        .filter(u -> !SymbolV2Utils.isFunctionOrClassDeclaration(u))
+        .map(UsageV2::tree)
+        .filter(NameImpl.class::isInstance)
+        .map(NameImpl.class::cast)
+        .forEach(n -> n.typeV2(rhsType));
```

Comment:
this part is the same as usages stream below, so maybe it makes sense to extend this common part?